File: ongoing/backend/fastapi_app/utils/error_handler.py
"""
エラーハンドリングの統一ユーティリティ
"""
from fastapi import HTTPException, Request
from fastapi.responses import JSONResponse
from typing import Dict, Any, Optional, Union
import logging
import traceback
from datetime import datetime

# ログ設定
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class APIError(Exception):
    """カスタムAPIエラークラス"""
    def __init__(self, message: str, status_code: int = 500, details: Optional[Dict[str, Any]] = None, error_code: Optional[str] = None):
        self.message = message
        self.status_code = status_code
        self.details = details or {}
        self.error_code = error_code
        super().__init__(self.message)

class ValidationError(APIError):
    """バリデーションエラー"""
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 400, details, "VALIDATION_ERROR")

class AuthenticationError(APIError):
    """認証エラー"""
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 401, details, "AUTHENTICATION_ERROR")

class AuthorizationError(APIError):
    """認可エラー"""
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 403, details, "AUTHORIZATION_ERROR")

class NotFoundError(APIError):
    """リソース未発見エラー"""
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 404, details, "NOT_FOUND_ERROR")

class ConflictError(APIError):
    """競合エラー"""
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 409, details, "CONFLICT_ERROR")

class ServiceUnavailableError(APIError):
    """サービス利用不可エラー"""
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 503, details, "SERVICE_UNAVAILABLE_ERROR")

def handle_api_error(error: Exception, operation: str) -> HTTPException:
    """APIエラーを統一形式で処理"""
    logger.error(f"❌ {operation}エラー: {error}")
    
    if isinstance(error, APIError):
        return HTTPException(
            status_code=error.status_code,
            detail={
                "message": error.message,
                "operation": operation,
                "error_code": error.error_code,
                "details": error.details,
                "timestamp": datetime.now().isoformat()
            }
        )
    elif isinstance(error, HTTPException):
        return error
    else:
        return HTTPException(
            status_code=500,
            detail={
                "message": f"{operation}中に予期しないエラーが発生しました",
                "operation": operation,
                "error_code": "INTERNAL_ERROR",
                "error": str(error),
                "timestamp": datetime.now().isoformat()
            }
        )

def create_success_response(data: Any, message: str = "Success", operation: str = "operation") -> Dict[str, Any]:
    """成功レスポンスの統一形式"""
    return {
        "status": "success",
        "message": message,
        "operation": operation,
        "data": data,
        "timestamp": datetime.now().isoformat()
    }

def create_error_response(message: str, operation: str, error_code: str = "UNKNOWN_ERROR", details: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """エラーレスポンスの統一形式"""
    return {
        "status": "error",
        "message": message,
        "operation": operation,
        "error_code": error_code,
        "details": details or {},
        "timestamp": datetime.now().isoformat()
    }

async def global_exception_handler(request: Request, exc: Exception) -> JSONResponse:
    """グローバル例外ハンドラー"""
    logger.error(f"❌ グローバル例外: {exc}")
    logger.error(f"❌ リクエストURL: {request.url}")
    logger.error(f"❌ スタックトレース: {traceback.format_exc()}")
    
    if isinstance(exc, APIError):
        return JSONResponse(
            status_code=exc.status_code,
            content=create_error_response(
                message=exc.message,
                operation="global_handler",
                error_code=exc.error_code or "API_ERROR",
                details=exc.details
            )
        )
    elif isinstance(exc, HTTPException):
        return JSONResponse(
            status_code=exc.status_code,
            content=create_error_response(
                message=exc.detail.get("message", str(exc.detail)) if isinstance(exc.detail, dict) else str(exc.detail),
                operation="global_handler",
                error_code="HTTP_ERROR"
            )
        )
    else:
        return JSONResponse(
            status_code=500,
            content=create_error_response(
                message="内部サーバーエラーが発生しました",
                operation="global_handler",
                error_code="INTERNAL_ERROR",
                details={"original_error": str(exc)}
            )
        )

def log_error(error: Exception, operation: str, context: Optional[Dict[str, Any]] = None):
    """エラーログの統一出力"""
    logger.error(f"❌ {operation}エラー: {error}")
    if context:
        logger.error(f"❌ コンテキスト: {context}")
    logger.error(f"❌ スタックトレース: {traceback.format_exc()}")

def validate_required_fields(data: Dict[str, Any], required_fields: list[str]) -> None:
    """必須フィールドのバリデーション"""
    missing_fields = [field for field in required_fields if field not in data or data[field] is None or data[field] == ""]
    if missing_fields:
        raise ValidationError(
            message=f"必須フィールドが不足しています: {', '.join(missing_fields)}",
            details={"missing_fields": missing_fields}
        )

def validate_http_status(response, operation: str) -> None:
    """HTTPレスポンスステータスのバリデーション"""
    if not response.ok:
        if response.status == 401:
            raise AuthenticationError(f"{operation}の認証に失敗しました")
        elif response.status == 403:
            raise AuthorizationError(f"{operation}の認可に失敗しました")
        elif response.status == 404:
            raise NotFoundError(f"{operation}のリソースが見つかりません")
        elif response.status == 409:
            raise ConflictError(f"{operation}で競合が発生しました")
        elif response.status >= 500:
            raise ServiceUnavailableError(f"{operation}のサービスが利用できません")
        else:
            raise APIError(f"{operation}でHTTPエラーが発生しました: {response.status}")

File: blog/20251111/frontend/next-app/loghoi/lib/errorHandler.ts
/**
 * 統一エラーハンドリングユーティリティ
 */

export interface APIErrorResponse {
  status: 'error' | 'success'
  message: string
  operation: string
  error_code?: string
  details?: Record<string, any>
  timestamp: string
}

export interface APISuccessResponse<T = any> {
  status: 'success'
  message: string
  operation: string
  data: T
  timestamp: string
}

export class APIError extends Error {
  public statusCode: number
  public errorCode?: string
  public details?: Record<string, any>
  public operation?: string

  constructor(
    message: string,
    statusCode: number = 500,
    errorCode?: string,
    details?: Record<string, any>,
    operation?: string
  ) {
    super(message)
    this.name = 'APIError'
    this.statusCode = statusCode
    this.errorCode = errorCode
    this.details = details
    this.operation = operation
  }
}

export class ValidationError extends APIError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 400, 'VALIDATION_ERROR', details, 'validation')
  }
}

export class AuthenticationError extends APIError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 401, 'AUTHENTICATION_ERROR', details, 'authentication')
  }
}

export class AuthorizationError extends APIError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 403, 'AUTHORIZATION_ERROR', details, 'authorization')
  }
}

export class NotFoundError extends APIError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 404, 'NOT_FOUND_ERROR', details, 'not_found')
  }
}

export class ConflictError extends APIError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 409, 'CONFLICT_ERROR', details, 'conflict')
  }
}

export class ServiceUnavailableError extends APIError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 503, 'SERVICE_UNAVAILABLE_ERROR', details, 'service_unavailable')
  }
}

/**
 * HTTPレスポンスを解析してエラーを判定
 */
export function parseAPIResponse<T = any>(
  response: Response,
  operation: string
): Promise<APISuccessResponse<T>> {
  return response.json().then((data: APIErrorResponse | APISuccessResponse<T>) => {
    if (data.status === 'error') {
      const errorData = data as APIErrorResponse
      throw createErrorFromResponse(errorData, operation)
    }
    return data as APISuccessResponse<T>
  })
}

/**
 * APIエラーレスポンスから適切なエラークラスを作成
 */
export function createErrorFromResponse(errorData: APIErrorResponse, operation: string): APIError {
  const { error_code, message, details } = errorData

  switch (error_code) {
    case 'VALIDATION_ERROR':
      return new ValidationError(message, details)
    case 'AUTHENTICATION_ERROR':
      return new AuthenticationError(message, details)
    case 'AUTHORIZATION_ERROR':
      return new AuthorizationError(message, details)
    case 'NOT_FOUND_ERROR':
      return new NotFoundError(message, details)
    case 'CONFLICT_ERROR':
      return new ConflictError(message, details)
    case 'SERVICE_UNAVAILABLE_ERROR':
      return new ServiceUnavailableError(message, details)
    default:
      return new APIError(message, 500, error_code, details, operation)
  }
}

/**
 * ユーザーフレンドリーなエラーメッセージを生成
 */
export function getUserFriendlyMessage(error: Error): string {
  if (error instanceof ValidationError) {
    return `入力内容に問題があります: ${error.message}`
  }
  
  if (error instanceof AuthenticationError) {
    return `認証に失敗しました: ${error.message}`
  }
  
  if (error instanceof AuthorizationError) {
    return `アクセス権限がありません: ${error.message}`
  }
  
  if (error instanceof NotFoundError) {
    return `データが見つかりません: ${error.message}`
  }
  
  if (error instanceof ConflictError) {
    return `操作が競合しています: ${error.message}`
  }
  
  if (error instanceof ServiceUnavailableError) {
    return `サービスが一時的に利用できません: ${error.message}`
  }
  
  if (error instanceof APIError) {
    return `APIエラー: ${error.message}`
  }
  
  // ネットワークエラーやその他のエラー
  if (error.name === 'TypeError' && error.message.includes('fetch')) {
    return 'ネットワーク接続に問題があります。インターネット接続を確認してください。'
  }
  
  if (error.name === 'AbortError') {
    return 'リクエストがタイムアウトしました。しばらく待ってから再試行してください。'
  }
  
  return `予期しないエラーが発生しました: ${error.message}`
}

/**
 * エラーログを出力
 */
export function logError(error: Error, context?: Record<string, any>): void {
  console.error('❌ エラーが発生しました:', {
    name: error.name,
    message: error.message,
    stack: error.stack,
    context,
    timestamp: new Date().toISOString()
  })
}

/**
 * エラーをユーザーに表示するためのアラートを生成
 */
export function createErrorAlert(error: Error): {
  type: 'error' | 'warning' | 'info'
  title: string
  message: string
  details?: string
} {
  const userMessage = getUserFriendlyMessage(error)
  
  if (error instanceof ValidationError) {
    return {
      type: 'warning',
      title: '入力エラー',
      message: userMessage,
      details: error.details ? JSON.stringify(error.details, null, 2) : undefined
    }
  }
  
  if (error instanceof AuthenticationError) {
    return {
      type: 'error',
      title: '認証エラー',
      message: userMessage
    }
  }
  
  if (error instanceof AuthorizationError) {
    return {
      type: 'error',
      title: 'アクセス権限エラー',
      message: userMessage
    }
  }
  
  if (error instanceof NotFoundError) {
    return {
      type: 'warning',
      title: 'データが見つかりません',
      message: userMessage
    }
  }
  
  if (error instanceof ServiceUnavailableError) {
    return {
      type: 'error',
      title: 'サービス利用不可',
      message: userMessage
    }
  }
  
  return {
    type: 'error',
    title: 'エラー',
    message: userMessage,
    details: error.stack
  }
}

/**
 * 必須フィールドのバリデーション
 */
export function validateRequiredFields(
  data: Record<string, any>,
  requiredFields: string[]
): void {
  const missingFields = requiredFields.filter(
    field => !data[field] || data[field] === ''
  )
  
  if (missingFields.length > 0) {
    throw new ValidationError(
      `必須フィールドが不足しています: ${missingFields.join(', ')}`,
      { missingFields }
    )
  }
}

/**
 * HTTPステータスコードからエラークラスを判定
 */
export function createErrorFromStatus(
  status: number,
  message: string,
  operation: string
): APIError {
  switch (status) {
    case 400:
      return new ValidationError(message)
    case 401:
      return new AuthenticationError(message)
    case 403:
      return new AuthorizationError(message)
    case 404:
      return new NotFoundError(message)
    case 409:
      return new ConflictError(message)
    case 503:
      return new ServiceUnavailableError(message)
    default:
      return new APIError(message, status, 'HTTP_ERROR', undefined, operation)
  }
}
